### DISCUSSION TITLE
Help with receiving RTK corrections

### DISCUSSION THREAD

**Discussion by vladimirlabynko** [semuconsulting/pygnssutils#47] [Q&A] [answered]
Hello. 
I found your lib, it's pretty cool and useful. Help me, please. I need to receive RTK corrections and display it in terminal I find code that can help me with that, but I receive coordinates without corrections. 
I add quality = parsed_data.quality in code that you showed in [issue10](https://github.com/semuconsulting/pygnssutils/issues/10) to see which GNSS Quality Indicator I. As I read in documentation from my receiver when it RTK received I should  see quality=5(recalculated coordinates). But I always receive 1(clean coordinates without RTK fix). What do I need to change in the code to get coordinates using RTK?

````
from sys import platform
from io import BufferedReader
from threading import Thread, Lock, Event
from queue import Queue
from time import sleep
from serial import Serial
from pyubx2 import (
    UBXReader,
    NMEA_PROTOCOL,
    UBX_PROTOCOL,
    RTCM3_PROTOCOL,
    protocol,
)
from pyrtcm import RTCM_MSGIDS
from pygnssutils import GNSSNTRIPClient, VERBOSITY_LOW, haversine

# NTRIP server parameters - AMEND AS REQUIRED:
# Ideally, mountpoint should be <30 km from location.
NTRIP_SERVER = "server"
NTRIP_PORT = 20
MOUNTPOINT = "mountpoint"
NTRIP_USER = "user"
NTRIP_PASSWORD = "pwd"

# Fixed reference coordinates - AMEND AS REQUIRED:
REFLAT = 53.9
REFLON = 27.5
REFALT = 124
REFSEP = 0

# Set to True to print entire GNSS/NTRIP message rather than just identity
PRINT_FULL = False


def read_gnss(stream, lock, stopevent):
    """
    THREADED
    Reads and parses incoming GNSS data from receiver.
    """

    ubr = UBXReader(
        BufferedReader(serial),
        protfilter=(NMEA_PROTOCOL | UBX_PROTOCOL | RTCM3_PROTOCOL),
    )

    while not stopevent.is_set():
        try:
            if stream.in_waiting:
                lock.acquire()
                (raw_data, parsed_data) = ubr.read()  # pylint: disable=unused-variable
                lock.release()
                if parsed_data:

                    # START OF DEVIATION LOGGING ILLUSTRATION
                    # if message contains lat/lon, compare it to reference coordinates
                    # uses haversine formula to estimate spherical distance ('deviation')
                    # between the two sets of coordinates
                    if hasattr(parsed_data, "lat") and hasattr(parsed_data, "lon") and hasattr(parsed_data, "quality"):
                        lat = parsed_data.lat
                        lon = parsed_data.lon
                        quality = parsed_data.quality
                        print(
                            f"Receiver coordinates: {lat}, {lon}. {quality}"
                        )
                    # END OF DEVIATION LOGGING ILLUSTRATION

                    idy = parsed_data.identity
                    # if it's an RXM-RTCM message, show which RTCM3 message
                    # it's acknowledging and whether it's been used or not.""
                    if idy == "RXM-RTCM":
                        nty = (
                            f" - {parsed_data.msgType} "
                            f"{'Used' if parsed_data.msgUsed > 0 else 'Not used'}"
                        )
                    else:
                        nty = ""
                    if PRINT_FULL:
                        print(parsed_data)
                    #else:
                        #print(f"GNSS>> {idy}{nty}")
        except Exception as err:
            print(f"Something went wrong in read thread {err}")
            continue


def send_gnss(stream, lock, stopevent, inqueue):
    """
    THREADED
    Reads RTCM3 data from message queue and sends it to receiver.
    """

    while not stopevent.is_set():
        try:
            raw_data, parsed_data = inqueue.get()
            if protocol(raw_data) == RTCM3_PROTOCOL:
                if PRINT_FULL:
                    print(parsed_data)
                else:
                    print(
                        f"NTRIP>> {parsed_data.identity} {RTCM_MSGIDS[parsed_data.identity]}"
                    )
                lock.acquire()
                stream.write(raw_data)
                lock.release()
        except Exception as err:
            print(f"Something went wrong in send thread {err}")
            continue


def start_read_thread(stream, lock, stopevent):
    """
    Start read thread.
    """

    rth = Thread(
        target=read_gnss,
        args=(
            stream,
            lock,
            stopevent,
        ),
        daemon=True,
    )
    rth.start()
    return rth


def start_send_thread(stream, lock, stopevent, msgqueue):
    """
    Start send thread.
    """

    kth = Thread(
        target=send_gnss,
        args=(
            stream,
            lock,
            stopevent,
            msgqueue,
        ),
        daemon=True,
    )
    kth.start()
    return kth


if __name__ == "__main__":
    # pylint: disable=invalid-name

    # GNSS receiver serial port parameters - AMEND AS REQUIRED:
    SERIAL_PORT = "COM13"

    BAUDRATE = 115200
    TIMEOUT = 0.6
    GGAMODE = 0  # use fixed reference position (0 = use live position)
    GGAINT = 10  # -1 = do not send NMEA GGA sentences

    serial_lock = Lock()
    ntrip_queue = Queue()
    stop = Event()

    try:

        print(f"Opening serial port {SERIAL_PORT} @ {BAUDRATE}...\n")
        with Serial(SERIAL_PORT, BAUDRATE, timeout=TIMEOUT) as serial:

            stop.clear()
            print("Starting read thread...\n")
            gnss_thread = start_read_thread(serial, serial_lock, stop)

            print("Starting send thread...\n")
            rtk_thread = start_send_thread(serial, serial_lock, stop, ntrip_queue)

            print(f"Starting NTRIP client on {NTRIP_SERVER}:{NTRIP_PORT}...\n")
            with GNSSNTRIPClient(None, verbosity=VERBOSITY_LOW) as gnc:

                streaming = gnc.run(
                    server=NTRIP_SERVER,
                    port=NTRIP_PORT,
                    mountpoint=MOUNTPOINT,
                    user=NTRIP_USER,
                    password=NTRIP_PASSWORD,
                    reflat=REFLAT,
                    reflon=REFLON,
                    refalt=REFALT,
                    refsep=REFSEP,
                    ggamode=GGAMODE,
                    ggainterval=GGAINT,
                    output=ntrip_queue,
                )

                while streaming and not stop.is_set():  # run until user presses CTRL-C
                    sleep(1)
                sleep(1)

    except KeyboardInterrupt:
        stop.set()

    print("Terminated by user")
````

**Comment by semuadmin** [comment]
Hi @vladimirlabynko 

"What do I need to change in the code to get coordinates using RTK?"

I'm afraid that's a difficult question to answer without knowing your specific setup:

1) What receiver are you using? ZED-F9P?
2) Which NTRIP caster are you using?
3) Does that NTRIP service offer a mountpoint within, say, 30km of your location?
4) Are you using the appropriate login credentials for that caster and mountpoint?
5) Are you able to verify that you're actually receiving NTRIP data from the mountpoint - you can check this using the `gnssntripclient`CLI utility e.g. 

```
gnssntripclient -S yourcaster.com -M YOURMOUNTPOINT --ggamode 1 --reflat 12.345 --reflon 12.345 --refalt 123 --refsep 123 --ntripuser yourusername@yourdomain.com --ntrippassword yourpassword
```

If you have the correct settings for your NTRIP caster, this should produce a stream of RTCM data.

Confirm this first and, once you're established that you are able to receive NTRIP data, get back to me with any further queries.

FYI I've converted the Issue to a Discussion Q&A topic as it's more a general request for information.

**Reply by vladimirlabynko** [reply]
Thank you for fast reply.          
1. I use TAU 1308
2.  I can't answer this, I don't know
3. Yes mountpoint is 20km near me.
4. Yes 
5. I checked it with vendor's software and get fixed coordinates, with CLI I can check only on Wednesday.

**Reply by semuadmin** [reply]
The TAU 1308 has a proprietary RTK implementation so you'll need to refer to the receiver's manufacturer for specific guidance. But assuming it is fully NTRIP 2.0 compliant, and not locked into the manufacturer's proprietary RTK software, the `pygnssutils` suite should work fine. But you will need to find out which NTRIP caster and port it uses as the `pygnssutils` NTRIP tools need this information - again, ask the manufacturer for advice.

**Comment by vladimirlabynko** [comment]
Conducted an experiment.  I connected through the software provided by the developers to the base station, after connection I started to receive coordinates with RTK correction. Then I disconnected from the software and started the script, for some time (about 3 minutes) I received data with correction, then again stopped receiving them and received "raw" without RTK.

**Comment by semuadmin** [comment]
I'm sorry @vladimirlabynko but I'm not in a position to provide support or guidance on third party software. I can endeavour to help with pygnssutils but ONLY if you are able to provide the NTRIP configuration information I have previously requested. 

I'm also unclear what you mean by "raw" in this context? The receiver should be outputting (_presumably_) 'raw' NMEA data (typically including GGA, GLL and GSA sentences) as a matter of course, irrespective of whether it's receiving RTK corrections or not. If the receiver is correctly receiving valid and usable RTCM data from the NTRIP caster, the GGA.quality flag would typically change from 1 or 2 (3D fix) to 4 (RTK FiXED) or 5 (RTK FLOAT). You should also start seeing non-blank values in the GGA.diffAge and GGA.diffStation fields - these refer to the ARP (Antenna Reference Point aka mountpoint) that is sending RTCM data.

What `script` are you actually running? - the one you've cut-and-pasted above was purely an illustration and would need editing for each individual user's NTRIP configuration.

**Comment by vladimirlabynko** [comment]
I run script that paste in my first message with my NTRIP settings(ip,port,login,password). All what I know about mountpoint that it used Leica GNSS Spider software.

**Comment by semuadmin** [comment, marked as answer]
> I run script that paste in my first message with my NTRIP settings(ip,port,login,password). All what I know about mountpoint that it used Leica GNSS Spider software.

Ok without knowing the NTRIP caster and mountpoint details I'm not sure there's much more I can do. As I say, I'm not in a position to advise on third party software. I'm 99% confident the script can be made to work with your receiver but you'll have to consult the manufacturer for details of the appropriate NTRIP configuration.
